### PR TITLE
Bosses

### PR DESCRIPTION
**Pull request by pmrkonjic**

### CODE REVIEW

File: src/api/java/com/github/minecraftschurlimods/arsmagicalegacy/api/entity/ISpellCasterEntity.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package com.github.minecraftschurlimods.arsmagicalegacy.api.entity;
+
+public interface ISpellCasterEntity {
```
Comment (Minecraftschurli) [inline src/api/java/com/github/minecraftschurlimods/arsmagicalegacy/api/entity/ISpellCasterEntity.java:6]: doc

File: src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/init/AMBlockEntities.java [hunk 2]
```diff
@@ -16,8 +16,8 @@
 
 @NonExtendable
 public interface AMBlockEntities {
-    RegistryObject<BlockEntityType<AltarCoreBlockEntity>>        ALTAR_CORE        = BLOCK_ENTITIES.register("altar_core",        () -> BlockEntityType.Builder.of(AltarCoreBlockEntity::new,        AMBlocks.ALTAR_CORE.get())       .build(null));
-    RegistryObject<BlockEntityType<AltarViewBlockEntity>>        ALTAR_VIEW        = BLOCK_ENTITIES.register("altar_view",        () -> BlockEntityType.Builder.of(AltarViewBlockEntity::new,        AMBlocks.ALTAR_VIEW.get())       .build(null));
+    RegistryObject<BlockEntityType<AltarCoreBlockEntity>> ALTAR_CORE = BLOCK_ENTITIES.register("altar_core", () -> BlockEntityType.Builder.of(AltarCoreBlockEntity::new, AMBlocks.ALTAR_CORE.get()).build(null));
+    RegistryObject<BlockEntityType<AltarViewBlockEntity>> ALTAR_VIEW = BLOCK_ENTITIES.register("altar_view",        () -> BlockEntityType.Builder.of(AltarViewBlockEntity::new,        AMBlocks.ALTAR_VIEW.get())       .build(null));
```
Comment (Minecraftschurli) [inline src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/init/AMBlockEntities.java]: why?

File: src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/init/AMBlocks.java [hunk 3]
```diff
@@ -53,8 +54,8 @@
 public interface AMBlocks {
     BlockBehaviour.Properties FLOWER_POT = BlockBehaviour.Properties.of(Material.DECORATION).instabreak().noOcclusion();
 
-    RegistryObject<OcculusBlock>          OCCULUS                  = BLOCKS.register("occulus",                  OcculusBlock::new);
-    RegistryObject<InscriptionTableBlock> INSCRIPTION_TABLE        = BLOCKS.register("inscription_table",        InscriptionTableBlock::new);
+    RegistryObject<OcculusBlock>          OCCULUS           = BLOCKS.register("occulus", OcculusBlock::new);
+    RegistryObject<InscriptionTableBlock> INSCRIPTION_TABLE = BLOCKS.register("inscription_table",        InscriptionTableBlock::new);
```
Comment (Minecraftschurli) [inline src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/init/AMBlocks.java]: why?

**Review by Minecraftschurli — CHANGES_REQUESTED** [review]
initial review

File: src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/init/AMBlocks.java (hunk 3 above)
Comment (IchHabeHunger54) [reply src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/init/AMBlocks.java]: intellidumb(TM)

File: src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/init/AMBlockEntities.java (hunk 2 above)
Comment (IchHabeHunger54) [reply src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/init/AMBlockEntities.java]: intellidumb(TM)

**Review by IchHabeHunger54 — DISMISSED** [review]

File: src/api/java/com/github/minecraftschurlimods/arsmagicalegacy/api/entity/ExecuteSpellGoal.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
+package com.github.minecraftschurlimods.arsmagicalegacy.api.entity;
+
+import com.github.minecraftschurlimods.arsmagicalegacy.api.spell.ISpell;
+import com.github.minecraftschurlimods.arsmagicalegacy.api.spell.SpellCastResult;
+import net.minecraft.sounds.SoundSource;
+import net.minecraft.world.entity.Mob;
+import net.minecraft.world.entity.ai.goal.Goal;
+
+public class ExecuteSpellGoal<T extends Mob & ISpellCasterEntity> extends Goal {
```
Comment (Minecraftschurli) [inline src/api/java/com/github/minecraftschurlimods/arsmagicalegacy/api/entity/ExecuteSpellGoal.java:9]: doc

File: src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/entity/EarthGuardian.java [hunk 5]
```diff
@@ -17,13 +22,14 @@
...
+    private EarthGuardianAction action;
 
     public EarthGuardian(EntityType<? extends EarthGuardian> type, Level level) {
-        super(type, level, BossEvent.BossBarColor.BLUE);
+        super(type, level, BossEvent.BossBarColor.GREEN);
     }
 
     public static AttributeSupplier.Builder createAttributes() {
-        return createMonsterAttributes().add(Attributes.FOLLOW_RANGE, Attributes.FOLLOW_RANGE.getDefaultValue()).add(Attributes.MAX_HEALTH, 140).add(Attributes.ARMOR, 23);
+        return createMonsterAttributes().add(Attributes.MAX_HEALTH, 120).add(Attributes.ARMOR, 20);
```
Comment (Minecraftschurli) [inline src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/entity/EarthGuardian.java]: missing mana and burnout attribute

File: src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/entity/EnderGuardian.java [hunk 6]
```diff
@@ -1,32 +1,44 @@
...
-
-    @Override
-    public float getEyeHeight(Pose pPose) {
-        return 2.5F;
-    }
-
-    @Override
-    protected void registerGoals() {
-        super.registerGoals();
+        return createMonsterAttributes().add(Attributes.MAX_HEALTH, 500).add(Attributes.ARMOR, 20);
```
Comment (Minecraftschurli) [inline src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/entity/EnderGuardian.java]: missing attributes

File: src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/entity/FireGuardian.java [hunk 7]
```diff
@@ -1,25 +1,38 @@
...
+    private FireGuardianAction action;
 
     public FireGuardian(EntityType<? extends FireGuardian> type, Level level) {
         super(type, level, BossEvent.BossBarColor.RED);
-        fireImmune();
     }
 
     public static AttributeSupplier.Builder createAttributes() {
-        return createMonsterAttributes().add(Attributes.FOLLOW_RANGE, Attributes.FOLLOW_RANGE.getDefaultValue()).add(Attributes.MAX_HEALTH, 250D).add(Attributes.ARMOR, 17);
+        return createMonsterAttributes().add(Attributes.MAX_HEALTH, 250).add(Attributes.ARMOR, 20);
```
Comment (Minecraftschurli) [inline src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/entity/FireGuardian.java]: again missing attributes

File: src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/init/AMEntities.java [hunk 8]
```diff
@@ -25,8 +25,8 @@
 import static com.github.minecraftschurlimods.arsmagicalegacy.common.init.AMRegistries.ENTITIES;
 
 public interface AMEntities {
-    RegistryObject<EntityType<Projectile>>        PROJECTILE         = ENTITIES.register("projectile",         () -> EntityType.Builder.of(Projectile::new,        MobCategory.MISC)   .clientTrackingRange(8).sized(0.25F, 0.25F).build("projectile"));
-    RegistryObject<EntityType<Wall>>              WALL               = ENTITIES.register("wall",               () -> EntityType.Builder.of(Wall::new,              MobCategory.MISC)   .clientTrackingRange(8).sized(0.25F, 0.25F).build("wall"));
+    RegistryObject<EntityType<Projectile>> PROJECTILE = ENTITIES.register("projectile", () -> EntityType.Builder.of(Projectile::new, MobCategory.MISC).clientTrackingRange(8).sized(0.25F, 0.25F).build("projectile"));
+    RegistryObject<EntityType<Wall>>       WALL       = ENTITIES.register("wall",               () -> EntityType.Builder.of(Wall::new,              MobCategory.MISC)   .clientTrackingRange(8).sized(0.25F, 0.25F).build("wall"));
```
Comment (Minecraftschurli) [inline src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/init/AMEntities.java]: why?

File: src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/init/AMItems.java [hunk 9]
```diff
@@ -44,77 +45,93 @@ public ItemStack makeIcon() {
...
-    RegistryObject<MageArmorItem>                BATTLEMAGE_HELMET        = ITEMS.register("battlemage_helmet", () -> new MageArmorItem(MageArmorItem.BATTLEMAGE_ARMOR_MATERIAL, EquipmentSlot.HEAD, 10));
-    RegistryObject<MageArmorItem>                BATTLEMAGE_CHESTPLATE    = ITEMS.register("battlemage_chestplate", () -> new MageArmorItem(MageArmorItem.BATTLEMAGE_ARMOR_MATERIAL, EquipmentSlot.CHEST, 10));
-    RegistryObject<MageArmorItem>                BATTLEMAGE_LEGGINGS      = ITEMS.register("battlemage_leggings", () -> new MageArmorItem(MageArmorItem.BATTLEMAGE_ARMOR_MATERIAL, EquipmentSlot.LEGS, 10));
-    RegistryObject<MageArmorItem>                BATTLEMAGE_BOOTS         = ITEMS.register("battlemage_boots", () -> new MageArmorItem(MageArmorItem.BATTLEMAGE_ARMOR_MATERIAL, EquipmentSlot.FEET, 10));
+    RegistryObject<InfinityOrbItem> INFINITY_ORB = ITEMS.register("infinity_orb", InfinityOrbItem::new);
+    RegistryObject<BlockItem>       OCCULUS      = registerBlockItem64(AMBlocks.OCCULUS);
```
Comment (Minecraftschurli) [inline src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/init/AMItems.java]: why?

**Review by Minecraftschurli — CHANGES_REQUESTED** [review]
Some things still to do

**Review by Minecraftschurli — APPROVED** [review]
Looks good :shipit:

**Pull request merged**